### PR TITLE
Xeps/bind2 inlines

### PR DESCRIPTION
**Pull request by NelsonVides**
This PR introduces BIND2 support for carbons, csi, and sm.

As described in
- https://xmpp.org/extensions/xep-0388.html (SASL2 XEP)
- https://xmpp.org/extensions/xep-0386.html (BIND2 XEP)
- https://github.com/xsf/xeps/pull/1215 (changes to the stream-management XEP specifying the interactions with SASL2 and BIND2)

Note that the changes to SM are not accepted yet (at the moment of writing), and for a while they claimed the SM feature inlined into SASL2 should be announced as `resume` but then it was requested it should announce `sm`, as other XMPP servers and clients have already done.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/189980) / elasticsearch_and_cassandra_mnesia / c4ef17da1637f5c840c7fb493aab3c37ef1b9361
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/189980/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/189980/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40fa7cb4ebbb70.2023-09-04_14.33.53/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190113) / elasticsearch_and_cassandra_mnesia / 8e9958a66015f423c59b8a57e5f1910e0ca4b8c9
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/190113/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/190113/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%401e200fe95eed.2023-09-05_07.58.08/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190177) / elasticsearch_and_cassandra_mnesia / 14a62f5214b152737e2e2a591276973bc63bd3a8
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/190177/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/190177/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40f1e3f68df2f7.2023-09-05_11.20.47/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190293) / elasticsearch_and_cassandra_mnesia / 41f35a9a3cd63fe1443ed0a24d608028f01b622e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/190293/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/190293/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40c183d7d94bf1.2023-09-06_07.39.24/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190313) / elasticsearch_and_cassandra_mnesia / 6829950f992b99e5a07c075c18b1b5dfbe69a670
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/190313/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/190313/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%406b9d7bbbde01.2023-09-06_08.45.42/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190359) / elasticsearch_and_cassandra_mnesia / b0904ed6cc4c8a444332ae34973d44fae26af7e2
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/190359/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/190359/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40c535df12253f.2023-09-06_09.06.01/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190434) / elasticsearch_and_cassandra_mnesia / 9e1ef30c5513e5de53233b3710525cc03357e20c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/190434/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/190434/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%4046feb0307445.2023-09-06_14.04.20/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190609) / elasticsearch_and_cassandra_mnesia / fe9e2ca65544d1d0bc4aedbdcdddcb141acd5b11
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/190609/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/190609/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40dc551fd67750.2023-09-07_13.24.36/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190804) / elasticsearch_and_cassandra_mnesia / b0b887e68efa02b2a91c55aea4ca7739cfb27652
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4114/190804/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4114/190804/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40df30e829d399.2023-09-11_08.06.15/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Review by chrzaszcz — APPROVED** [review]
Looks good :+1:

File: src/stream_management/mod_stream_management_sasl2.erl [hunk 1]
```diff
@@ -28,7 +28,7 @@ hooks(HostType) ->
 -spec sasl2_stream_features(Acc, #{c2s_data := mongoose_c2s:data()}, gen_hook:extra()) ->
     {ok, Acc} when Acc :: [exml:element()].
 sasl2_stream_features(Acc, _, _) ->
-    Resume = #xmlel{name = <<"resume">>, attrs = [{<<"xmlns">>, ?NS_STREAM_MGNT_3}]},
+    Resume = #xmlel{name = <<"sm">>, attrs = [{<<"xmlns">>, ?NS_STREAM_MGNT_3}]},
```
Comment (chrzaszcz) [inline src/stream_management/mod_stream_management_sasl2.erl:31]: Where is this change described in the XEP? It seems to be against https://xmpp.org/extensions/xep-0388.html#example-2

Comment (NelsonVides) [reply src/stream_management/mod_stream_management_sasl2.erl:31]: That's a bug in the XEP that I found testing with Conversations. Conversations and Prosody already implemented using <sm/> and teh XSF group discussed changing the XEP to use such implementation instead, see https://github.com/xsf/xeps/pull/1215/files#r1321210647

**Comment by Neustradamus** [comment]
@NelsonVides: Good job!

**Pull request merged**